File: PSME/application-ref/src/rest-ref/endpoints/ethernet_switch.cpp
/*!
 * @copyright
 * Copyright (c) 2015-2016 Intel Corporation
 *
 * @copyright
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * @copyright
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * @copyright
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * */

#include "psme/rest-ref/endpoints/ethernet_switch.hpp"
#include "psme/rest-ref/constants/constants.hpp"
#include "json/json.hpp"

using namespace psme::rest;
using namespace app::rest::constants;



namespace {
json::Value make_prototype() {
    json::Value r(json::Value::Type::OBJECT);

    r[Common::ODATA_CONTEXT] =
        "/redfish/v1/$metadata#EthernetSwitches/Members/$entity";
    r[Common::ODATA_ID] = json::Value::Type::NIL;
    r[Common::ODATA_TYPE] = "#EthernetSwitch.1.0.0.EthernetSwitch";
    r[Common::ID] = json::Value::Type::NIL;
    r[Common::NAME] = "Ethernet Switch";
    r[Common::DESCRIPTION] = "Ethernet Switch description";
    r[Switch::SWITCH_ID] = json::Value::Type::NIL;
    r[Common::MANUFACTURER] = json::Value::Type::NIL;
    r[Common::MODEL] = json::Value::Type::NIL;
    r[Switch::MANUFACTURING_DATE] = json::Value::Type::NIL;
    r[Common::SERIAL] = json::Value::Type::NIL;
    r[Common::PART_NUMBER] = json::Value::Type::NIL;
    r[Switch::FIRMWARE_NAME] = json::Value::Type::NIL;
    r[Common::FIRMWARE_VERSION] = json::Value::Type::NIL;
    r[Switch::ROLE] = json::Value::Type::NIL;
    r[Switch::MAX_ACL_NUMBER] = json::Value::Type::NIL;

    r[Common::STATUS][Common::STATE] = json::Value::Type::NIL;
    r[Common::STATUS][Common::HEALTH] = json::Value::Type::NIL;
    r[Common::STATUS][Common::HEALTH_ROLLUP] = json::Value::Type::NIL;

    r[Common::OEM] = json::Value::Type::OBJECT;

    r[Switch::PORTS] = json::Value::Type::OBJECT;
    r[Switch::ACLS] = json::Value::Type::OBJECT;
    r[Common::LINKS][Switch::CHASSIS] = json::Value::Type::NIL;
    r[Common::LINKS][Common::MANAGED_BY] = json::Value::Type::ARRAY;
    r[Common::LINKS][Common::OEM] = json::Value::Type::OBJECT;

    return r;
}

}

endpoint::EthernetSwitch::EthernetSwitch(const std::string& path) : Endpoint(path) {}
endpoint::EthernetSwitch::~EthernetSwitch() {}

void endpoint::EthernetSwitch::get(const server::Request& req, server::Response& res) {
    auto r = ::make_prototype();

    auto s =  psme::rest::model::Find<agent_framework::model::Switch>
        (req.params[PathParam::ETHERNET_SWITCH_ID]).get();

    r[Common::ODATA_ID] = PathBuilder(req).build();
    r[Common::ID] = req.params[PathParam::ETHERNET_SWITCH_ID];
    r[Common::NAME] = Switch::SWITCH +
        req.params[PathParam::ETHERNET_SWITCH_ID];
    r[Switch::SWITCH_ID] = s.get_switch_identifier();

    const auto& fru = s.get_fru_info();
    r[Common::MANUFACTURER] = fru.get_manufacturer();
    r[Common::MODEL] = fru.get_model_number();
    r[Switch::MANUFACTURING_DATE] = s.get_manufacturing_date();
    r[Common::SERIAL] = fru.get_serial_number();
    r[Common::PART_NUMBER] = fru.get_part_number();
    r[Switch::FIRMWARE_NAME] = s.get_firmware_name();
    r[Common::FIRMWARE_VERSION] = s.get_firmware_version();
    r[Switch::ROLE] = s.get_role();
    psme::rest::endpoint::utils::status_to_json(s,r);
    r[Switch::MAX_ACL_NUMBER] = s.get_max_acl_number();
    r[Common::STATUS][Common::HEALTH_ROLLUP] =
        endpoint::utils::rollup_switch_health(s.get_uuid()).to_string();
    r[Switch::PORTS][Common::ODATA_ID] = PathBuilder(req).append(
        Switch::PORTS).build();
    r[Switch::ACLS][Common::ODATA_ID] = PathBuilder(req).append(
        Switch::ACLS).build();

    r[Common::LINKS]
     [Switch::CHASSIS]
     [Common::ODATA_ID] = PathBuilder(PathParam::BASE_URL)
        .append(Root::CHASSIS)
        .append(utils::get_chassis_by_uuid(s.get_chassis()).get_id()).build();

    json::Value manager;
    manager[Common::ODATA_ID] = PathBuilder(PathParam::BASE_URL)
        .append(Root::MANAGERS)
        .append(utils::get_manager_by_uuid(s.get_parent_uuid()).get_id()).build();

    r[Common::LINKS]
     [Common::MANAGED_BY]
     [Collection::MEMBERS].push_back(std::move(manager));
    set_response(res, r);
}

File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/EthernetSwitch.java
/*
 * Copyright (c) 2016-2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.Eventable;
import com.intel.podm.business.entities.SuppressEvents;
import com.intel.podm.business.entities.redfish.base.DiscoverableEntity;
import com.intel.podm.business.entities.redfish.base.Entity;
import com.intel.podm.common.types.Id;

import javax.persistence.Column;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.NamedQueries;
import javax.persistence.NamedQuery;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static com.intel.podm.common.utils.Contracts.requiresNonNull;
import static javax.persistence.CascadeType.MERGE;
import static javax.persistence.CascadeType.PERSIST;
import static javax.persistence.FetchType.LAZY;

@javax.persistence.Entity
@NamedQueries({
    @NamedQuery(name = EthernetSwitch.GET_ALL_ETHERNET_SWITCH_IDS,
        query = "SELECT ethernetSwitch.entityId FROM EthernetSwitch ethernetSwitch")
})
@Table(name = "ethernet_switch", indexes = @Index(name = "idx_ethernet_switch_entity_id", columnList = "entity_id", unique = true))
@Eventable
@SuppressWarnings({"checkstyle:MethodCount"})
public class EthernetSwitch extends DiscoverableEntity {
    public static final String GET_ALL_ETHERNET_SWITCH_IDS = "GET_ALL_ETHERNET_SWITCH_IDS";

    @Column(name = "entity_id", columnDefinition = ENTITY_ID_STRING_COLUMN_DEFINITION)
    private Id entityId;

    @Column(name = "switch_id")
    private String switchId;

    @Column(name = "manufacturer")
    private String manufacturer;

    @Column(name = "manufacturing_date")
    private String manufacturingDate;

    @Column(name = "model")
    private String model;

    @Column(name = "serial_number")
    private String serialNumber;

    @Column(name = "part_number")
    private String partNumber;

    @Column(name = "firmware_name")
    private String firmwareName;

    @Column(name = "firmware_version")
    private String firmwareVersion;

    @Column(name = "role")
    private String role;

    @SuppressEvents
    @OneToMany(mappedBy = "ethernetSwitch", fetch = LAZY, cascade = {MERGE, PERSIST})
    private Set<EthernetSwitchPort> ports = new HashSet<>();

    @SuppressEvents
    @OneToMany(mappedBy = "ethernetSwitch", fetch = LAZY, cascade = {MERGE, PERSIST})
    private Set<EthernetSwitchAcl> acls = new HashSet<>();

    @ManyToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "ethernet_switch_metrics_id")
    private EthernetSwitchMetrics ethernetSwitchMetrics;

    @ManyToMany(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinTable(
        name = "ethernet_switch_manager",
        joinColumns = {@JoinColumn(name = "ethernet_switch_id", referencedColumnName = "id")},
        inverseJoinColumns = {@JoinColumn(name = "manager_id", referencedColumnName = "id")})
    private Set<Manager> managers = new HashSet<>();

    @ManyToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "chassis_id")
    private Chassis chassis;

    @Override
    public Id getId() {
        return entityId;
    }

    @Override
    public void setId(Id id) {
        entityId = id;
    }

    public String getSwitchId() {
        return switchId;
    }

    public void setSwitchId(String switchId) {
        this.switchId = switchId;
    }

    public String getManufacturer() {
        return manufacturer;
    }

    public void setManufacturer(String manufacturer) {
        this.manufacturer = manufacturer;
    }

    public String getManufacturingDate() {
        return manufacturingDate;
    }

    public void setManufacturingDate(String manufacturingDate) {
        this.manufacturingDate = manufacturingDate;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public String getSerialNumber() {
        return serialNumber;
    }

    public void setSerialNumber(String serialNumber) {
        this.serialNumber = serialNumber;
    }

    public String getPartNumber() {
        return partNumber;
    }

    public void setPartNumber(String partNumber) {
        this.partNumber = partNumber;
    }

    public String getFirmwareName() {
        return firmwareName;
    }

    public void setFirmwareName(String firmwareName) {
        this.firmwareName = firmwareName;
    }

    public String getFirmwareVersion() {
        return firmwareVersion;
    }

    public void setFirmwareVersion(String version) {
        this.firmwareVersion = version;
    }

    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        this.role = role;
    }

    public Set<EthernetSwitchPort> getPorts() {
        return ports;
    }

    public void addPort(EthernetSwitchPort ethernetSwitchPort) {
        requiresNonNull(ethernetSwitchPort, "ethernetSwitchPort");

        ports.add(ethernetSwitchPort);
        if (!this.equals(ethernetSwitchPort.getEthernetSwitch())) {
            ethernetSwitchPort.setEthernetSwitch(this);
        }
    }

    public void unlinkPort(EthernetSwitchPort ethernetSwitchPort) {
        if (ports.contains(ethernetSwitchPort)) {
            ports.remove(ethernetSwitchPort);
            if (ethernetSwitchPort != null) {
                ethernetSwitchPort.unlinkEthernetSwitch(this);
            }
        }
    }

    public EthernetSwitchMetrics getEthernetSwitchMetrics() {
        return ethernetSwitchMetrics;
    }

    public void setSwitchMetrics(EthernetSwitchMetrics ethernetSwitchMetrics) {
        if (!Objects.equals(this.ethernetSwitchMetrics, ethernetSwitchMetrics)) {
            unlinkEthernetSwitchMetrics(this.ethernetSwitchMetrics);
            this.ethernetSwitchMetrics = ethernetSwitchMetrics;
            if (ethernetSwitchMetrics != null && !this.equals(ethernetSwitchMetrics.getEthernetSwitch())) {
                ethernetSwitchMetrics.setEthernetSwitch(this);
            }
        }
    }

    public void unlinkEthernetSwitchMetrics(EthernetSwitchMetrics ethernetSwitchMetrics) {
        if (Objects.equals(this.ethernetSwitchMetrics, ethernetSwitchMetrics)) {
            this.ethernetSwitchMetrics = null;
            if (ethernetSwitchMetrics != null) {
                ethernetSwitchMetrics.unlinkEthernetSwitch(this);
            }
        }
    }

    public Set<EthernetSwitchAcl> getAcls() {
        return acls;
    }

    public void addAcl(EthernetSwitchAcl ethernetSwitchAcl) {
        requiresNonNull(ethernetSwitchAcl, "ethernetSwitchAcl");

        acls.add(ethernetSwitchAcl);
        if (!this.equals(ethernetSwitchAcl.getEthernetSwitch())) {
            ethernetSwitchAcl.setEthernetSwitch(this);
        }
    }

    public void unlinkAcl(EthernetSwitchAcl ethernetSwitchAcl) {
        if (acls.contains(ethernetSwitchAcl)) {
            acls.remove(ethernetSwitchAcl);
            if (ethernetSwitchAcl != null) {
                ethernetSwitchAcl.unlinkEthernetSwitch(this);
            }
        }
    }

    public Set<Manager> getManagers() {
        return managers;
    }

    public void addManager(Manager manager) {
        requiresNonNull(manager, "manager");

        managers.add(manager);
        if (!manager.getEthernetSwitches().contains(this)) {
            manager.addEthernetSwitch(this);
        }
    }

    public void unlinkManager(Manager manager) {
        if (managers.contains(manager)) {
            managers.remove(manager);
            if (manager != null) {
                manager.unlinkEthernetSwitch(this);
            }
        }
    }

    public Chassis getChassis() {
        return chassis;
    }

    public void setChassis(Chassis chassis) {
        if (!Objects.equals(this.chassis, chassis)) {
            unlinkChassis(this.chassis);
            this.chassis = chassis;
            if (chassis != null && !chassis.getEthernetSwitches().contains(this)) {
                chassis.addEthernetSwitch(this);
            }
        }
    }

    public void unlinkChassis(Chassis chassis) {
        if (Objects.equals(this.chassis, chassis)) {
            this.chassis = null;
            if (chassis != null) {
                chassis.unlinkEthernetSwitch(this);
            }
        }
    }

    @Override
    public void preRemove() {
        unlinkCollection(ports, this::unlinkPort);
        unlinkCollection(acls, this::unlinkAcl);
        unlinkCollection(managers, this::unlinkManager);
        unlinkChassis(chassis);
        unlinkEthernetSwitchMetrics(ethernetSwitchMetrics);
    }

    @Override
    public boolean containedBy(Entity possibleParent) {
        return isContainedBy(possibleParent, chassis);
    }
}
